File: go/pray_calc.go
// Package praycalc provides Islamic prayer time calculations.
// Pure Go, no dependencies.
package praycalc

import (
	"fmt"
	"math"
	"time"
)

// Method represents an Islamic prayer time calculation method.
type Method string

const (
	MethodISNA      Method = "isna"
	MethodMWL       Method = "mwl"
	MethodEgypt     Method = "egypt"
	MethodUmmAlQura Method = "umm_al_qura"
	MethodTehran    Method = "tehran"
	MethodKarachi   Method = "karachi"
)

// Madhab determines the Asr prayer calculation method.
type Madhab string

const (
	MadhabShafi  Madhab = "shafi"
	MadhabHanafi Madhab = "hanafi"
)

// PrayerTimes holds the calculated prayer times for a day.
type PrayerTimes struct {
	Fajr      string  // HH:MM 24-hour
	Sunrise   string
	Dhuhr     string
	Asr       string
	Maghrib   string
	Isha      string
	Date      string  // YYYY-MM-DD
	Latitude  float64
	Longitude float64
	Method    Method
	Madhab    Madhab
}

type methodParams struct {
	fajrAngle    float64
	ishaAngle    float64 // 0 if using interval
	ishaInterval int     // minutes after maghrib; 0 if using angle
}

var methods = map[Method]methodParams{
	MethodISNA:      {15.0, 15.0, 0},
	MethodMWL:       {18.0, 17.0, 0},
	MethodEgypt:     {19.5, 17.5, 0},
	MethodUmmAlQura: {18.5, 0, 90},
	MethodTehran:    {17.7, 14.0, 0},
	MethodKarachi:   {18.0, 18.0, 0},
}

func toRad(d float64) float64 { return d * math.Pi / 180 }
func toDeg(r float64) float64 { return r * 180 / math.Pi }

func fixHour(h float64) float64 {
	return h - 24*math.Floor(h/24)
}

func julianDay(year, month, day int) float64 {
	if month <= 2 {
		year--
		month += 12
	}
	a := math.Floor(float64(year) / 100)
	b := 2 - a + math.Floor(a/4)
	return math.Floor(365.25*float64(year+4716)) + math.Floor(30.6001*float64(month+1)) + float64(day) + b - 1524.5
}

func sunPosition(jd float64) (dec, eqTime float64) {
	d := jd - 2451545.0
	g := toRad(357.529 + 0.98560028*d)
	q := 280.459 + 0.98564736*d
	l := toRad(q + 1.915*math.Sin(g) + 0.020*math.Sin(2*g))
	e := toRad(23.439 - 0.00000036*d)
	ra := toDeg(math.Atan2(math.Cos(e)*math.Sin(l), math.Cos(l))) / 15
	ra = fixHour(ra)
	dec = toDeg(math.Asin(math.Sin(e) * math.Sin(l)))
	eqTime = q/15 - ra
	return
}

func sunAngleTime(jd, lat, angle float64, clockwise bool) float64 {
	dec, eqTime := sunPosition(jd)
	noon := 12.0 - eqTime
	cosVal := (-math.Sin(toRad(angle)) - math.Sin(toRad(dec))*math.Sin(toRad(lat))) /
		(math.Cos(toRad(dec)) * math.Cos(toRad(lat)))
	if cosVal < -1 || cosVal > 1 {
		return math.NaN()
	}
	t := toDeg(math.Acos(cosVal)) / 15
	if clockwise {
		return noon + t
	}
	return noon - t
}

func asrTime(jd, lat float64, shadowFactor int) float64 {
	dec, eqTime := sunPosition(jd)
	noon := 12.0 - eqTime
	targetAngle := -toDeg(math.Atan(1.0 / (float64(shadowFactor) + math.Tan(math.Abs(toRad(lat-dec))))))
	cosVal := (-math.Sin(toRad(targetAngle)) - math.Sin(toRad(dec))*math.Sin(toRad(lat))) /
		(math.Cos(toRad(dec)) * math.Cos(toRad(lat)))
	if cosVal < -1 || cosVal > 1 {
		return math.NaN()
	}
	return noon + toDeg(math.Acos(cosVal))/15
}

func formatTime(hour, lng float64) string {
	hour = fixHour(hour + lng/15)
	h := int(hour)
	m := int((hour-float64(h))*60 + 0.5)
	if m >= 60 {
		h++
		m -= 60
	}
	return fmt.Sprintf("%02d:%02d", h%24, m)
}

// Calculate computes Islamic prayer times for the given location and date.
func Calculate(lat, lng float64, d time.Time, method Method, madhab Madhab) PrayerTimes {
	params, ok := methods[method]
	if !ok {
		params = methods[MethodISNA]
		method = MethodISNA
	}

	jd := julianDay(d.Year(), int(d.Month()), d.Day())
	shadowFactor := 1
	if madhab == MadhabHanafi {
		shadowFactor = 2
	}

	fajr := sunAngleTime(jd, lat, params.fajrAngle, false)
	sunrise := sunAngleTime(jd, lat, -0.833, false)
	_, eqTime := sunPosition(jd)
	dhuhr := 12.0 - eqTime
	asr := asrTime(jd, lat, shadowFactor)
	maghrib := sunAngleTime(jd, lat, -0.833, true)

	var isha float64
	if params.ishaInterval > 0 {
		isha = maghrib + float64(params.ishaInterval)/60
	} else {
		isha = sunAngleTime(jd, lat, params.ishaAngle, true)
	}

	return PrayerTimes{
		Fajr:      formatTime(fajr, lng),
		Sunrise:   formatTime(sunrise, lng),
		Dhuhr:     formatTime(dhuhr, lng),
		Asr:       formatTime(asr, lng),
		Maghrib:   formatTime(maghrib, lng),
		Isha:      formatTime(isha, lng),
		Date:      d.Format("2006-01-02"),
		Latitude:  lat,
		Longitude: lng,
		Method:    method,
		Madhab:    madhab,
	}
}

File: swift/Sources/PrayCalc/PrayCalc.swift
/// Islamic prayer time calculation library for Swift.
/// Pure Swift, no dependencies.
import Foundation

/// Calculation method for prayer times.
public enum Method: String, CaseIterable {
    case isna = "isna"
    case mwl = "mwl"
    case egypt = "egypt"
    case ummAlQura = "umm_al_qura"
    case tehran = "tehran"
    case karachi = "karachi"
}

/// Madhab (school of jurisprudence) for Asr time calculation.
public enum Madhab: String {
    case shafi = "shafi"
    case hanafi = "hanafi"
}

/// Calculated prayer times for a given day and location.
public struct PrayerTimes {
    public let fajr: String
    public let sunrise: String
    public let dhuhr: String
    public let asr: String
    public let maghrib: String
    public let isha: String
    public let date: String
    public let latitude: Double
    public let longitude: Double
    public let method: Method
    public let madhab: Madhab
}

private struct MethodParams {
    let fajrAngle: Double
    let ishaAngle: Double    // 0 if using interval
    let ishaInterval: Int    // minutes after maghrib; 0 if using angle
}

private let methodParams: [Method: MethodParams] = [
    .isna:       MethodParams(fajrAngle: 15.0, ishaAngle: 15.0, ishaInterval: 0),
    .mwl:        MethodParams(fajrAngle: 18.0, ishaAngle: 17.0, ishaInterval: 0),
    .egypt:      MethodParams(fajrAngle: 19.5, ishaAngle: 17.5, ishaInterval: 0),
    .ummAlQura:  MethodParams(fajrAngle: 18.5, ishaAngle: 0,    ishaInterval: 90),
    .tehran:     MethodParams(fajrAngle: 17.7, ishaAngle: 14.0, ishaInterval: 0),
    .karachi:    MethodParams(fajrAngle: 18.0, ishaAngle: 18.0, ishaInterval: 0),
]

private func toRad(_ d: Double) -> Double { d * .pi / 180 }
private func toDeg(_ r: Double) -> Double { r * 180 / .pi }
private func fixHour(_ h: Double) -> Double { h - 24 * floor(h / 24) }

private func julianDay(year: Int, month: Int, day: Int) -> Double {
    var y = year, m = month
    if m <= 2 { y -= 1; m += 12 }
    let a = floor(Double(y) / 100)
    let b = 2 - a + floor(a / 4)
    return floor(365.25 * Double(y + 4716)) + floor(30.6001 * Double(m + 1)) + Double(day) + b - 1524.5
}

private func sunPosition(jd: Double) -> (dec: Double, eqTime: Double) {
    let d = jd - 2451545.0
    let g = toRad(357.529 + 0.98560028 * d)
    let q = 280.459 + 0.98564736 * d
    let l = toRad(q + 1.915 * sin(g) + 0.020 * sin(2 * g))
    let e = toRad(23.439 - 0.00000036 * d)
    var ra = toDeg(atan2(cos(e) * sin(l), cos(l))) / 15
    ra = fixHour(ra)
    let dec = toDeg(asin(sin(e) * sin(l)))
    let eqTime = q / 15 - ra
    return (dec, eqTime)
}

private func sunAngleTime(jd: Double, lat: Double, angle: Double, clockwise: Bool) -> Double {
    let (dec, eqTime) = sunPosition(jd: jd)
    let noon = 12.0 - eqTime
    let cosVal = (-sin(toRad(angle)) - sin(toRad(dec)) * sin(toRad(lat))) /
                 (cos(toRad(dec)) * cos(toRad(lat)))
    guard cosVal >= -1, cosVal <= 1 else { return Double.nan }
    let t = toDeg(acos(cosVal)) / 15
    return clockwise ? noon + t : noon - t
}

private func asrTime(jd: Double, lat: Double, shadowFactor: Int) -> Double {
    let (dec, eqTime) = sunPosition(jd: jd)
    let noon = 12.0 - eqTime
    let targetAngle = -toDeg(atan(1.0 / (Double(shadowFactor) + tan(abs(toRad(lat - dec))))))
    let cosVal = (-sin(toRad(targetAngle)) - sin(toRad(dec)) * sin(toRad(lat))) /
                 (cos(toRad(dec)) * cos(toRad(lat)))
    guard cosVal >= -1, cosVal <= 1 else { return Double.nan }
    return noon + toDeg(acos(cosVal)) / 15
}

/// Rendered in place of a prayer that has no time on the requested date.
///
/// Above the polar circles the sun can fail to rise or set for weeks, so a depression
/// angle is simply unreachable and `sunAngleTime`/`asrTime` return NaN.
public let noTimePlaceholder = "--:--"

private func formatTime(_ hour: Double, lng: Double) -> String {
    // `Int(Double.nan)` is a FATAL TRAP in Swift, not a catchable error: it killed the
    // host process outright for any caller asking about a polar latitude (PKG-06).
    guard hour.isFinite else { return noTimePlaceholder }
    let h = fixHour(hour + lng / 15)
    guard h.isFinite else { return noTimePlaceholder }
    var hr = Int(h)
    var min = Int((h - Double(hr)) * 60 + 0.5)
    if min >= 60 { hr += 1; min -= 60 }
    return String(format: "%02d:%02d", hr % 24, min)
}

/// Calculate Islamic prayer times for a given location and date.
/// - Parameters:
///   - latitude: Decimal degrees (-90 to 90)
///   - longitude: Decimal degrees (-180 to 180)
///   - date: The date to calculate for
///   - method: Calculation method (default: .isna)
///   - madhab: Asr shadow ratio (default: .shafi)
/// - Returns: PrayerTimes with formatted HH:MM times
public func calculatePrayerTimes(
    latitude: Double,
    longitude: Double,
    date: Date,
    method: Method = .isna,
    madhab: Madhab = .shafi
) -> PrayerTimes {
    let cal = Calendar(identifier: .gregorian)
    let comps = cal.dateComponents([.year, .month, .day], from: date)
    let jd = julianDay(year: comps.year!, month: comps.month!, day: comps.day!)
    let params = methodParams[method]!
    let shadowFactor = madhab == .shafi ? 1 : 2

    let fajr = sunAngleTime(jd: jd, lat: latitude, angle: params.fajrAngle, clockwise: false)
    let sunrise = sunAngleTime(jd: jd, lat: latitude, angle: -0.833, clockwise: false)
    let (_, eqTime) = sunPosition(jd: jd)
    let dhuhr = 12.0 - eqTime
    let asr = asrTime(jd: jd, lat: latitude, shadowFactor: shadowFactor)
    let maghrib = sunAngleTime(jd: jd, lat: latitude, angle: -0.833, clockwise: true)
    let isha: Double = params.ishaInterval > 0
        ? maghrib + Double(params.ishaInterval) / 60
        : sunAngleTime(jd: jd, lat: latitude, angle: params.ishaAngle, clockwise: true)

    let df = DateFormatter()
    df.dateFormat = "yyyy-MM-dd"

    return PrayerTimes(
        fajr: formatTime(fajr, lng: longitude),
        sunrise: formatTime(sunrise, lng: longitude),
        dhuhr: formatTime(dhuhr, lng: longitude),
        asr: formatTime(asr, lng: longitude),
        maghrib: formatTime(maghrib, lng: longitude),
        isha: formatTime(isha, lng: longitude),
        date: df.string(from: date),
        latitude: latitude,
        longitude: longitude,
        method: method,
        madhab: madhab
    )
}
